subprojects/core/src/integTest/groovy/org/gradle/api/tasks/FailingIncrementalTasksIntegrationTest.groovy

```diff
@@ -33,15 +34,83 @@ class FailingIncrementalTasksIntegrationTest extends AbstractIntegrationSpec {
             }
         """
 
+        expect:
+        succeeds "foo"
+
         when:
-        run "foo"
         file("out.txt") << "force rerun"
-        def failure1 = runAndFail "foo"
-        def failure2 = runAndFail "foo"
+        fails"foo"
+        then:
+        failureHasCause "Boo!"
 
+        when:
+        fails "foo", "--info"
         then:
-        failure1.assertHasCause("Boo!")
-        failure2.assertHasCause("Boo!")
+        failureHasCause "Boo!"
+        output.contains "Task has failed previously."
         //this exposes an issue we used to have with in-memory cache.
     }
+
+    @Unroll
+    def "incremental task after previous #failiureCount failure(s) #description"() {
```

Comment:
I found the test: https://github.com/gradle/gradle/blob/4093a70098e6ab86d29013b6ee8091f2f135db2f/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalTasksIntegrationTest.groovy#L427-L427. Sorry about the noise.